### PR TITLE
Add "components" docs

### PR DESCRIPTION
**Pull request by wpscholar**
## Description
Autogenerate components documentation. Pulls in all readme files from https://github.com/WordPress/gutenberg/tree/master/packages/components and automatically adds them to the Gutenberg docs. Implemented `lint-staged` and `husky` pre-commit hook to automatically run a build of the docs in the event that a file has changed that will likely result in a need for rebuilding the docs.  Additionally, I automated validation of package.json files and linting of JS files that have changed.

Fixes https://github.com/WordPress/gutenberg/issues/4952

## How has this been tested?
I've manually tested the scripts and can see the appropriate entries added to the manifest.json file. I've also tested the `lint-staged` pre-commit hooks to ensure that they work properly.

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by gziolo** [comment]
Awesome work. I will try to review this PR later this week 💯 

By the way, @mmtr is working on improvements for the documentation of components. Both #8267 and #8338 should be good PRs to test together with those changes.

File: package.json
Comment (gziolo) [inline package.json]: We should rather use `npm run lint-pkg-json` in here to ensure it always uses the universal command.

File: package.json
Comment (gziolo) [inline package.json]: We should rather use `npm run lint-js` in here to ensure it always uses the universal command.

File: package.json
Comment (gziolo) [inline package.json]: I bet this should include all type of `README.md` files located inside `packages` folder.
We should also update when new `packages/*/README.md` is added.

Comment (ntwb) [reply package.json]: Is it overkill to run this on every commit? How long does this task take to run?

Comment (gziolo) [reply package.json]: I think it only generates `manifest.json` file. It's super fast.

**Review by ntwb — APPROVED** [review]

File: package.json
Comment (wpscholar) [reply package.json]: @gziolo If we do that, then it will run for all of the package.json files and not just the ones that have changed.

File: package.json
Comment (wpscholar) [reply package.json]: @gziolo If we do that, then it will run for all of the JavaScript files and not just the ones that have changed. It is very slow when scanning everything.

File: package.json
Comment (wpscholar) [reply package.json]: @ntwb It only runs if a file has changed that matches the patterns listed here. It is also super fast. 

@gziolo Agreed.

File: package.json
Comment (wpscholar) [reply package.json]: Essentially the lint staged tool will automatically pass the files that have changed to the command. Because the `npm run lint-js` script passes `.` as the path, it will cause it to run `eslint` against all `.js` files in the project. By just using the `eslint` command here, only the files that have changed are linted. Besides, we don't want linting problems caused by someone else's commit to prevent our own.

Comment (ntwb) [reply package.json]: Unless you use `git add -p` for partially staging files, if you do stage files in this manner `lint-staged` is unable to detect the difference.

See https://github.com/okonet/lint-staged/issues/62

**Comment by wpscholar** [comment]
@gziolo I've pushed an update to the PR... would love to get a final review from you. Should I also wait on the other reviewers you added before we merge?

File: docs/tool/config.js
Comment (youknowriad) [inline docs/tool/config.js]: This is meant to be a "config" file, I'd personally move all the logic to generate the manifest to `docs/tool/manifest.js` and leave here only the parts that are "configurable"

Comment (youknowriad) [reply docs/tool/config.js]: I guess the line is not always easy to draw. We should probably rethink this config globally though. So don't take this comment as a big blocker.

Comment (wpscholar) [reply docs/tool/config.js]: That is what I was going to do initially, but that file exports a single function that is very specific to the logic for the packages and data docs. I considered exporting an object from that file which contained more than one function but went the path of least resistance. I'd be happy to make some changes if you have a specific recommendation.

Comment (youknowriad) [reply docs/tool/config.js]: Since the function is already doing two things, it can do three :) and we can split it into three functions if necessary.

I'd only keep `'packages/components/src'` as a config variable. Something like: `componentsSource` or something like that, and just move everything else to `manifest.js`.

Also it looks like the `baseRepoUrl` is already being used in that file (hardcoded), so it might be good to use this variable consistently.

File: package.json
Comment (gziolo) [reply package.json]: _Aside:_ we need `wp-scripts lint-js` anyways, that would be a way to provide one interface in the long run.

File: docs/tool/config.js
Comment (gziolo) [reply docs/tool/config.js]: Let's address this feedback and 🚢 

I really like the automation proposed in this PR 💯

**Comment by wpscholar** [comment]
@youknowriad If you can give this a final review, I'd appreciate it. Thanks!

File: docs/tool/manifest.js
Comment (youknowriad) [inline docs/tool/manifest.js]: Minor; I would have moved those functions above (and added JSDocs) and only export the `getManifest`concatenating the three manifests.

**Review by youknowriad — APPROVED** [review]
Nice work

**Comment by gziolo** [comment]
Applied two more commits to include the latest generated version of files.

**Comment by gziolo** [comment]
Another merge conflict resolved. I will merge it as soon as Travis is happy.

Great work @wpscholar. Let's iterate on it in the follow-up PRs.

I think we should add support in pre-commit hook for `stylelint` which was added today by @noisysocks.

**Pull request merged**
